src/protocols/secure_channel/CASESession.cpp

```diff
@@ -683,64 +667,42 @@ CHIP_ERROR CASESession::RecoverInitiatorIpk()
+void CASESession::HandleConnectionAttemptComplete(Transport::ActiveTCPConnectionHolder & conn, CHIP_ERROR err)
 {
-    VerifyOrReturn(conn != nullptr);
-    // conn->mAppState should not be NULL. SessionManager has already checked
-    // before calling this callback.
-    VerifyOrDie(conn->mAppState != nullptr);
+    VerifyOrReturn(conn == mPeerConnState);
 
     char peerAddrBuf[chip::Transport::PeerAddress::kMaxToStringSize];
     conn->mPeerAddr.ToString(peerAddrBuf);
 
-    CASESession * caseSession = reinterpret_cast<CASESession *>(conn->mAppState->appContext);
-    VerifyOrReturn(caseSession != nullptr);
+    auto ConnectionCleanup = [](CASESession * s) {
+        s->mExchangeCtxt.Value()->GetSessionHandle()->AsUnauthenticatedSession()->ReleaseTCPConnection();
+        s->mSecureSessionHolder.Get().Value()->AsSecureSession()->ReleaseTCPConnection();
+        s->Clear();
+    };
+    std::unique_ptr<CASESession, decltype(ConnectionCleanup)> connectionHolder(this, ConnectionCleanup);
 
-    // Exit and disconnect if connection setup encountered an error.
-    SuccessOrExit(err);
+    // Bail if connection setup encountered an error.
+    LogAndReturnOnFailure(err, SecureChannel, "Connection establishment failed with peer at %s", peerAddrBuf);
 
     ChipLogDetail(SecureChannel, "TCP Connection established with %s before session establishment", peerAddrBuf);
 
     // Associate the connection with the current unauthenticated session for the
     // CASE exchange.
-    caseSession->mExchangeCtxt.Value()->GetSessionHandle()->AsUnauthenticatedSession()->SetTCPConnection(conn);
+    mExchangeCtxt.Value()->GetSessionHandle()->AsUnauthenticatedSession()->SetTCPConnection(conn);
 
     // Associate the connection with the current secure session that is being
     // set up.
-    caseSession->mSecureSessionHolder.Get().Value()->AsSecureSession()->SetTCPConnection(conn);
+    mSecureSessionHolder.Get().Value()->AsSecureSession()->SetTCPConnection(conn);
 
     // Send Sigma1 after connection is established for sessions over TCP
-    err = caseSession->SendSigma1();
-    SuccessOrExit(err);
+    LogAndReturnOnFailure(SendSigma1(), SecureChannel, "Sigma1 failed to peer %s", peerAddrBuf);
 
-exit:
-    if (err != CHIP_NO_ERROR)
-    {
-        ChipLogError(SecureChannel, "Connection establishment failed with peer at %s: %" CHIP_ERROR_FORMAT, peerAddrBuf,
-                     err.Format());
-
-        // Close the underlying connection and ensure that the CASESession is
-        // not holding on to a stale ActiveTCPConnectionState. We call
-        // TCPDisconnect() here explicitly in order to abort the connection
-        // even after it establishes successfully, but SendSigma1() fails for
-        // some reason.
-        caseSession->mSessionManager->TCPDisconnect(conn, /* shouldAbort = */ true);
-        caseSession->mPeerConnState = nullptr;
-
-        caseSession->Clear();
-    }
+    connectionHolder.release();
```

Comment:
So naming-wise: connectionHolder is not really holding a connection.  It's conceptually holding an instance of connnection auto-cleanup, and releasing it means not doing the cleanup, right?  

Might be better named connectionAutoCleanup.